Optimize updating key by storing lease in lessor

nekto0n opened the pull request:
This is a first attempt to avoid looking up `KeyValue` in storage and unmarshalling protobuf  mentioned in #6615. Also optimized `delete`, which had to do the same.
Here is somewhat strange `mvcc` package benchmark comparison (do not look at timings, run benchmarks on my laptop):
```
benchmark                                   old ns/op     new ns/op      delta
BenchmarkIndexRestore-4                     7795          7809           +0.18%
BenchmarkStorePut-4                         42041         42061          +0.05%
BenchmarkStorePutUpdate-4                   61554         42147          -31.53%
BenchmarkStoreTxnPut-4                      43972         44419          +1.02%
BenchmarkWatchableStorePut-4                42125         41386          -1.75%
BenchmarkWatchableStoreTxnPut-4             49307         53009          +7.51%
BenchmarkWatchableStoreWatchSyncPut-4       7520          7376           -1.91%
BenchmarkWatchableStoreUnsyncedCancel-4     908           1792           +97.36%
BenchmarkWatchableStoreSyncedCancel-4       0.82          1187249426     +144786515265.85%
BenchmarkKVWatcherMemoryUsage-4             24495         23046          -5.92%

benchmark                                   old allocs     new allocs     delta
BenchmarkIndexRestore-4                     3              3              +0.00%
BenchmarkStorePut-4                         12             12             +0.00%
BenchmarkStorePutUpdate-4                   18             12             -33.33%
BenchmarkStoreTxnPut-4                      12             12             +0.00%
BenchmarkWatchableStorePut-4                11             11             +0.00%
BenchmarkWatchableStoreTxnPut-4             14             14             +0.00%
BenchmarkWatchableStoreWatchSyncPut-4       2              2              +0.00%
BenchmarkWatchableStoreUnsyncedCancel-4     0              0              +0.00%
BenchmarkWatchableStoreSyncedCancel-4       0              37             +Inf%
BenchmarkKVWatcherMemoryUsage-4             9              9              +0.00%

benchmark                                   old bytes     new bytes     delta
BenchmarkIndexRestore-4                     178           178           +0.00%
BenchmarkStorePut-4                         1541          1524          -1.10%
BenchmarkStorePutUpdate-4                   5148          3752          -27.12%
BenchmarkStoreTxnPut-4                      1545          1557          +0.78%
BenchmarkWatchableStorePut-4                1289          1262          -2.09%
BenchmarkWatchableStoreTxnPut-4             1333          1307          -1.95%
BenchmarkWatchableStoreWatchSyncPut-4       207           203           -1.93%
BenchmarkWatchableStoreUnsyncedCancel-4     0             0             +0.00%
BenchmarkWatchableStoreSyncedCancel-4       0             2352          +Inf%
BenchmarkKVWatcherMemoryUsage-4             547           545           -0.37%

```
Not quite sure why `BenchmarkWatchableStoreSyncedCancel` brings such results.

nekto0n commented:
There is another opportunity to avoid allocating new string to construct `LeaseItem` to perform lookup in `itemMap` (go optimizes  string construction away in `itemMap[string(key)])` if `key` is `[]bytes`), but I am not sure if it is okay to bring bytes to `Lessor`.

heyitsanthony commented on mvcc/kvstore_test.go:
```diff
@@ -144,7 +144,6 @@ func TestStorePut(t *testing.T) {
 
 		if tt.rr != nil {
 			wact = []testutil.Action{
-				{"range", []interface{}{keyBucketName, newTestKeyBytes(tt.r.rev, false), []byte(nil), int64(0)}},
```
why remove these?

nekto0n replied:
We don't lookup in storage to get `LeaseID` anymore, we use lessor for that.

heyitsanthony replied:
OK

nekto0n commented on lease/lessor.go:
```diff
@@ -76,6 +76,8 @@ type Lessor interface {
 	// If the lease does not exist, an error will be returned.
 	Attach(id LeaseID, items []LeaseItem) error
 
+	GetLease(item LeaseItem) LeaseID
```
Oh, right. Forgot it. Will do.

heyitsanthony commented:
lgtm /cc @xiang90

xiang90 commented:
Can we figure out why 

```
BenchmarkWatchableStoreSyncedCancel-4       0.82          1187249426     +144786515265.85%
```

before get this merged?

nekto0n commented on lease/lessor.go:
```diff
@@ -361,10 +368,15 @@ func (le *lessor) Attach(id LeaseID, items []LeaseItem) error {
 }
 
+func (le *lessor) GetLease(item LeaseItem) LeaseID {
+	return le.itemMap[item]
```
Right, we do. Fixed.

nekto0n commented:
@xiang90 I think `BenchmarkWatchableStoreSyncedCancel` is flaky. These I my runs one after another with this patch applied:
```
BenchmarkWatchableStoreSyncedCancel-4   	       1	1158563857 ns/op	    2352 B/op	      37 allocs/op
BenchmarkWatchableStoreSyncedCancel-4   	       1	1027656738 ns/op	    2048 B/op	      32 allocs/op
BenchmarkWatchableStoreSyncedCancel-4   	      50	  21017376 ns/op	     340 B/op	       5 allocs/op
```
And these are runs on `master`:
```
BenchmarkWatchableStoreSyncedCancel-4   	      10	 112671224 ns/op	    1099 B/op	      17 allocs/op
BenchmarkWatchableStoreSyncedCancel-4   	10000000	       100 ns/op	       0 B/op	       0 allocs/op
BenchmarkWatchableStoreSyncedCancel-4   	100000000	        10.0 ns/op	       0 B/op	       0 allocs/op
```
I've never seen such inconsistent things.

nekto0n commented:
Here cpuprofile from run which looks like this
```
BenchmarkWatchableStoreSyncedCancel-4   	   20000	     51045 ns/op	       1 B/op	       0 allocs/op
```

```
(pprof) top
58.48s of 81.18s total (72.04%)
Dropped 154 nodes (cum <= 0.41s)
Showing top 10 nodes out of 85 (cum >= 7.63s)
      flat  flat%   sum%        cum   cum%
    13.77s 16.96% 16.96%     24.67s 30.39%  runtime.scanobject
    12.04s 14.83% 31.79%     12.47s 15.36%  runtime.mapaccess2_fast64
     6.93s  8.54% 40.33%     10.28s 12.66%  runtime.evacuate
     6.57s  8.09% 48.42%     21.96s 27.05%  runtime.mapassign1
     6.14s  7.56% 55.99%      6.14s  7.56%  runtime.heapBitsForObject
     4.44s  5.47% 61.46%         5s  6.16%  runtime.greyobject
     3.33s  4.10% 65.56%      4.01s  4.94%  runtime.mapdelete
     1.82s  2.24% 67.80%      1.82s  2.24%  runtime.aeshash64
     1.73s  2.13% 69.93%      1.73s  2.13%  runtime.procyield
     1.71s  2.11% 72.04%      7.63s  9.40%  github.com/coreos/etcd/mvcc.(*watcherGroup).delete
```
Seems like GC is kicking in randomly. Running with `GOGC=off` gives a bit better results and here is profile

```
(pprof) top
20750ms of 28770ms total (72.12%)
Dropped 83 nodes (cum <= 143.85ms)
Showing top 10 nodes out of 72 (cum >= 610ms)
      flat  flat%   sum%        cum   cum%
    6350ms 22.07% 22.07%     6620ms 23.01%  runtime.mapaccess2_fast64
    3890ms 13.52% 35.59%     5420ms 18.84%  runtime.evacuate
    3670ms 12.76% 48.35%    10580ms 36.77%  runtime.mapassign1
    1830ms  6.36% 54.71%     2270ms  7.89%  runtime.mapdelete
    1170ms  4.07% 58.78%     1170ms  4.07%  runtime.aeshash64
     970ms  3.37% 62.15%     4680ms 16.27%  github.com/coreos/etcd/mvcc.(*watcherGroup).delete
     850ms  2.95% 65.10%    27640ms 96.07%  github.com/coreos/etcd/mvcc.BenchmarkWatchableStoreSyncedCancel
     770ms  2.68% 67.78%      770ms  2.68%  runtime.memmove
     640ms  2.22% 70.00%      640ms  2.22%  runtime.memclr
     610ms  2.12% 72.12%      610ms  2.12%  sync/atomic.AddUint32
```

nekto0n commented:
If I lower number of watchers to 100k I consistenly get:
```
BenchmarkWatchableStoreSyncedCancel-4   	2000000000	         0.04 ns/op	       0 B/op	       0 allocs/op
```

xiang90 commented:
LGTM
